server/src/main/java/org/jboss/as/server/deployment/module/ModuleDependency.java

```diff
@@ -91,6 +108,28 @@ public ModuleDependency(final ModuleLoader moduleLoader, final ModuleIdentifier
         this.moduleLoader = moduleLoader;
         this.importServices = importServices;
         this.userSpecified = userSpecified;
+        this.reason = Optional.empty();
+    }
+
+    /**
+     * Construct a new instance.
+     *
+     * @param moduleLoader the module loader of the dependency (if {@code null}, then use the default server module loader)
+     * @param identifier the module identifier
+     * @param optional {@code true} if this is an optional dependency
+     * @param export {@code true} if resources should be exported by default
+     * @param importServices
+     * @param userSpecified {@code true} if this dependency was specified by the user, {@code false} if it was automatically added
+     * @param reason reason for adding implicit module dependency
+     */
+    public ModuleDependency(final ModuleLoader moduleLoader, final ModuleIdentifier identifier, final boolean optional, final boolean export, final boolean importServices, final boolean userSpecified, Optional<String> reason) {
```

Comment:
Don't use an Optional<String> for the constructor and replace it with a String reason.

The callers don't need to an Optional, the callers will have two constructors, one with an additional argument for the reason, so by choosing one constructor instead of the other, they are already defining whether the reason is optional.

This makes it more concise avoiding callers the need to add think about if they should use Optional.empty() if there is no reason available.

Getting an Optional<String> for the getReason() makes sense for third parties.

Reply:
What do you mean by "it makes sense for third parties"? Can you please clarify this for me? 
Because there is this comment from Darran https://wildfly.zulipchat.com/#narrow/stream/174184-wildfly-developers/topic/Adding.20reasons.20behind.20implicit.20dependencies/near/313017577 saying that many DUPs exists outside of our codebase, and I am not sure now if this somehow touches this or not.

Reply:
It follows up on what we discussed here: https://github.com/wildfly/wildfly-core/pull/5353#discussion_r1085073266

I was a bit reluctant to use an Optional, but you gave an argument when a dependency can be used by any third party (third party feature packs and subsystems).

We have two constructors, so we should not use an Optional for the constructor since choosing one or the other will do the optional choice of given a reason. However, we could have third-party code that can invoke the getReason(), so it makes sense to me to leave it returning an Optional as discussed before.

Does it clarify your question?

Reply:
Yes, very much. :)
Thank you for clarifying.